### PR TITLE
Implement achievement cycles

### LINKED ISSUE #23: Achievement cycle
Every time an achievement is granted we should enter another cycle of achievement checking to ensure now enabled achievements are granted correctly.

### PR DESCRIPTION
**Pull request by frederike-ramin**
Feedback on the implementation is greatly appreciated. I don't think that's the prettiest way to do it, just couldn't come up with a different, better way.

Closes #23

### CODE REVIEW

File: test/config/achievement-rule-checker-config.yml [hunk 1]
```diff
@@ -52,8 +55,13 @@ achievements:
         requirements:
             - xp: XP
               amount: 10
-    AchievementCanBeAwadedTwice:
+    AchievementCanBeAwardedTwiceAtOnce:
```
Comment (corinnaj) [inline test/config/achievement-rule-checker-config.yml]: I think you can just take this out, it doesn't really make sense to configure the service in this manner.

File: test/hooks/achievement-rule-checker.test.js [hunk 2]
```diff
@@ -194,4 +194,59 @@ describe('\'achievement-rule-checker\' hook', () => {
 
     assert.deepEqual(result[0].amount, 2);
   });
+
+  it('gives maxAwarded achievements at once', async () => {
```
Comment (corinnaj) [inline test/hooks/achievement-rule-checker.test.js]: same here

File: src/rule-parser.js [hunk 3]
```diff
@@ -161,7 +161,11 @@ class EventRequirement extends  Requirement {
     });
   }
 
-  async isFulfilled(context) {
+  async isFulfilled(context, isFirstCycle) {
+    // event requirements should be only checked in first achievement cycle
+    if (!isFirstCycle) {
```
Comment (corinnaj) [inline src/rule-parser.js]: Is it possible that an achievement triggers an event, which is skipped here?

File: test/hooks/achievement-rule-checker.test.js (hunk 2 above)
Comment (frederike-ramin) [reply test/hooks/achievement-rule-checker.test.js]: In my opinion more tests don't do any harm, so I would probably just leave it in as it is wanted behavior, although not the standard case.

File: src/rule-parser.js (hunk 3 above)
Comment (frederike-ramin) [reply src/rule-parser.js]: Good point. Do we want to listen on events we produce? If so, we should probably give the user a possibility to trigger events in the action parts of configuration.

**Pull request merged**